### PR TITLE
Custom math classes

### PR DESCRIPTION
**Pull request by rqy2002**
This PR implement custom math classes.

Now it has several problems:
- [x] maybe use a new type instead of `i64` to store the class.
- [ ] docs.
- [ ] tests.

### CODE REVIEW

**Comment by rqy2002** [comment]
Sorry I'm not familiar with git and mess my branch up :cry:

**Comment by rqy2002** [comment]
I use `MathClass` instead.
Now there is a question: do we need other math classes, such as `Fence`, `Unary` and `Vary`?

File: library/src/math/class.rs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+/// ## Example
+/// ```example
+/// $ a operator(x o r) b, a x o r b $
+/// ```
+///
+/// Display: Math Operator
+/// Category: math
+/// Returns: content
+#[func]
+pub fn operator(
```
Comment (laurmaedje) [inline library/src/math/class.rs:16]: Do we really want these individual functions? Isn't `class("operator", ..)` fine?

File: library/src/math/class.rs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+) -> Value {
+    MathClassElem::new(MathClass::Punctuation, body).pack().into()
+}
+
+/// Math content with specified math class.
+///
+/// Display: Class
+/// Category: Math
+#[element(LayoutMath)]
+pub struct MathClassElem {
```
Comment (laurmaedje) [inline library/src/math/class.rs:138]: Since the function is called `class`, this should be called `ClassElem`.

File: src/eval/cast.rs [hunk 3]
```diff
@@ -181,6 +182,30 @@ cast_to_value! {
...
 
+cast_from_value! {
+    MathClass,
+    "operator" => MathClass::Large,
+    "opening" => MathClass::Opening,
+    "closing" => MathClass::Closing,
+    "binary" => MathClass::Binary,
+    "relation" => MathClass::Relation,
+    "normal" => MathClass::Normal,
+    "punctuation" => MathClass::Punctuation,
```
Comment (laurmaedje) [inline src/eval/cast.rs:193]: I think we should at least add fence, large, unary, and vary. Also, could you order this and the match below in the same order as [here](https://docs.rs/unicode-math-class/0.1.0/src/unicode_math_class/lib.rs.html#42-58).

File: src/eval/cast.rs [hunk 4]
```diff
@@ -181,6 +182,30 @@ cast_to_value! {
...
+cast_to_value! {
+    v: MathClass => Value::from(match v {
+        MathClass::Large => "operator",
+        MathClass::Opening => "opening",
+        MathClass::Closing => "closing",
+        MathClass::Binary => "binary",
+        MathClass::Relation => "relation",
+        MathClass::Normal => "normal",
+        MathClass::Punctuation => "punctuation",
+        _ => "unused math class",
```
Comment (laurmaedje) [inline src/eval/cast.rs:205]: This should either panic or just list all variants (even if they are not in the list above).

**Comment by laurmaedje** [comment]
Hey @rqy2002, are you still interested in completing this?

**Comment by rqy2002** [comment]
> Hey @rqy2002, are you still interested in completing this?

Maybe not. And I will be busy for a while and may not have enough time to do this. Sorry.

**Comment by laurmaedje** [comment]
No problem!

**Pull request closed without merging**